Firestore/core/src/firebase/firestore/util/path.cc

```diff
@@ -16,96 +16,225 @@
   }
 
   // Otherwise everything after the slash is the basename (even if empty string)
-  return pathname.substr(slash + 1);
+  size_t start = slash + 1;
+  return Path{pathname_.substr(start)};
 }
 
-absl::string_view Path::Dirname(absl::string_view pathname) {
-  size_t last_slash = pathname.find_last_of('/');
-
+Path Path::Dirname() const {
+  size_t last_slash = LastSeparator(c_str(), size());
   if (last_slash == npos) {
     // No path separator found => empty string. Conformance with POSIX would
     // have us return "." here.
-    return pathname.substr(0, 0);
+    return Path{string_type{}};
   }
 
   // Collapse runs of slashes.
-  size_t nonslash = pathname.find_last_not_of('/', last_slash);
-  if (nonslash == npos) {
+  size_t non_slash = LastNonSeparator(c_str(), last_slash);
+  if (non_slash == npos) {
     // All characters preceding the last path separator are slashes
-    return pathname.substr(0, 1);
+    return Path{pathname_.substr(0, 1)};
   }
 
-  last_slash = nonslash + 1;
-
   // Otherwise everything up to the slash is the parent directory
-  return pathname.substr(0, last_slash);
+  last_slash = non_slash + 1;
+  return Path{pathname_.substr(0, last_slash)};
+}
+
+bool Path::IsAbsolute() const {
+  return util::IsAbsolute(c_str(), size());
 }
 
-bool Path::IsAbsolute(absl::string_view path) {
-  path = StripDriveLetter(path);
-  return !path.empty() && IsSeparator(path.front());
+Path Path::AppendUtf8(absl::string_view path) const {
+  Path result{*this};
+  result.MutableAppendUtf8(path);
+  return result;
 }
 
-void Path::JoinAppend(std::string* base, absl::string_view path) {
-  if (IsAbsolute(path)) {
-    base->assign(path.data(), path.size());
+void Path::MutableAppendSegment(const char_type* path, size_t size) {
+  if (util::IsAbsolute(path, size)) {
+    pathname_.assign(path, size);
 
   } else {
-    size_t nonslash = base->find_last_not_of('/');
-    if (nonslash != npos) {
-      base->resize(nonslash + 1);
-      base->push_back('/');
+    size_t non_slash = LastNonSeparator(pathname_.c_str(), pathname_.size());
+    if (non_slash != npos) {
+      pathname_.resize(non_slash + 1);
+      pathname_.push_back(kPreferredSeparator);
     }
 
     // If path started with a slash we'd treat it as absolute above
-    base->append(path.data(), path.size());
+    pathname_.append(path, size);
   }
 }
 
+void Path::MutableAppendSegmentUtf8(absl::string_view path) {
+#if defined(_WIN32)
+  Path segment = Path::FromUtf8(path);
+  MutableAppendSegment(segment.c_str(), segment.size());
+
+#else
+  MutableAppendSegment(path.data(), path.size());
+#endif
+}
+
+#if defined(_WIN32)
+Path::string_type Path::CanonicalPath() const {
+  if (pathname_.empty()) {
+    return pathname_;
+  }
+
+  // Remove trailing separators, but take care not to remove the trailing slash
+  // in a root path name (which can have a drive letter).
+  std::wstring copy{pathname_};
+  size_t min = StripDriveLetter(copy.c_str(), copy.size());
```

Comment:
I couldn't understand what `min` stands for without reading the code. Perhaps something like `relative_path_begin`, `rel_path_first`, etc.?